pkg/cloud/errors.go

```diff
@@ -31,67 +30,42 @@ func IsNotFound(err error) bool {
 // IsRoleAssignmentAlreadyDeleted returns true if the given error is a role assignment already deleted error.
 // Ref: https://docs.microsoft.com/en-us/rest/api/authorization/role-assignments/delete#response
 func IsRoleAssignmentAlreadyDeleted(err error) bool {
-	derr := autorest.DetailedError{}
+	derr := &azcore.ResponseError{}
 	return errors.As(err, &derr) && derr.StatusCode == http.StatusNoContent
 }
 
-// IsAlreadyExists parses the error message to check if it's resource already exists error.
-func IsAlreadyExists(err error) bool {
-	derr := autorest.DetailedError{}
+// IsRoleAssignmentExists returns true if the given error is a role assignment already exists error.
+func IsRoleAssignmentExists(err error) bool {
+	derr := &azcore.ResponseError{}
 	return errors.As(err, &derr) && derr.StatusCode == http.StatusConflict
 }
 
 // IsFederatedCredentialNotFound returns true if the given error is a federated credential not found error.
 func IsFederatedCredentialNotFound(err error) bool {
 	gerr := GraphError{}
-	return errors.As(err, &gerr) && *gerr.PublicError.GetCode() == GraphErrorCodeResourceNotFound
+	return errors.As(err, &gerr) && *gerr.Errorable.GetCode() == GraphErrorCodeResourceNotFound
 }
 
 // IsFederatedCredentialAlreadyExists returns true if the given error is a federated credential already exists error.
 // E1202 22:40:05.500821  867104 main.go:57] "failed to add federated identity credential" err="code: Request_MultipleObjectsWithSameKeyValue, message: FederatedIdentityCredential with name aramase-default-cred already exists."
 func IsFederatedCredentialAlreadyExists(err error) bool {
 	gerr := GraphError{}
-	return errors.As(err, &gerr) && *gerr.PublicError.GetCode() == GraphErrorCodeMultipleObjectsWithSameKeyValue
+	return errors.As(err, &gerr) && *gerr.Errorable.GetCode() == GraphErrorCodeMultipleObjectsWithSameKeyValue
 }
 
 // GetGraphError returns the public error message from the additional info.
 // ref: https://docs.microsoft.com/en-us/graph/errors#error-resource-type
 // errors returned by the graph API aren't serialized today and this is a known issue: https://github.com/microsoftgraph/msgraph-sdk-go-core/issues/1
-func GetGraphError(additionalData map[string]interface{}) (*GraphError, error) {
-	if additionalData == nil || additionalData["error"] == nil {
-		return nil, nil
+func getGraphError(err error) *GraphError {
+	var oerr *odataerrors.ODataError
+	if errors.As(err, &oerr) {
+		return &GraphError{Errorable: oerr.GetError()}
```

Comment:
Based on how this function is used, something like this is probably easier to read (avoid the extra `nil` checks):

```suggestion
func maybeExtractGraphError(err error) error {
	var oerr *odataerrors.ODataError
	if errors.As(err, &oerr) {
		return &GraphError{Errorable: oerr.GetError()}
	}
	return err
```